### PR TITLE
feat(react): Introduce hooks

### PR DESCRIPTION
**Pull request by desiprisg**
### Screenshots

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

### CODE REVIEW

File: packages/react/src/components/NovuProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  const { children, ...options } = props;
+  const novu = useMemo(() => new Novu(options), []);
+
+  return <NovuContext.Provider value={novu}>{children}</NovuContext.Provider>;
+};
+
+export const useNovu = () => {
+  const context = useContext(NovuContext);
+  if (!context) {
+    throw new Error('useNovu must be used within a NovuProvider');
```
Comment (SokratisVidros) [inline packages/react/src/components/NovuProvider.tsx]:
```suggestion
    throw new Error('useNovu must be used within a <NovuProvider />');
```

File: packages/react/src/hooks/useNotifications.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+export type UseNotificationsResult = {
+  readAll: () => Promise<void>;
+  archiveAll: () => Promise<void>;
+  archiveAllRead: () => Promise<void>;
+  data?: Array<Notification>;
+  error?: NovuError;
+  isLoading: boolean;
+  isFetching: boolean;
+  refetch: () => Promise<void>;
+  fetchMore: () => void;
```
Comment (SokratisVidros) [inline packages/react/src/hooks/useNotifications.ts]: Isn't this a promise?

File: packages/react/src/hooks/useNotifications.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  archived?: boolean;
+  limit?: number;
+  onSuccess?: (data: Notification[]) => void;
+  onError?: (error: NovuError) => void;
+};
+
+export const useNotifications = (props?: UseNotificationsProps): UseNotificationsResult => {
+  const { tags, read, archived, limit, onSuccess, onError } = props || {};
+  const { notifications, on } = useNovu();
+  const [data, setData] = useState<Array<Notification>>();
```
Comment (SokratisVidros) [inline packages/react/src/hooks/useNotifications.ts:18]: @desiprisg This solution works with the internal useState works. Did you consider using https://react.dev/reference/react/useSyncExternalStore instead and make it listen to the Novu events? 

I feel like it might help us with the `setData` handling as it's a hook tailored to be used with underlying event-driven state managers like the Novu class.

Comment (desiprisg) [reply packages/react/src/hooks/useNotifications.ts:18]: will try it out

Comment (desiprisg) [reply packages/react/src/hooks/useNotifications.ts:18]: Tbh I like the other approach more, since we have to call list() anyway in a useEffect. It also acts as a more universal example for other framework implementations.

Comment (desiprisg) [reply packages/react/src/hooks/useNotifications.ts:18]: Marking as resolved as discussed offline

File: packages/react/src/hooks/index.ts [hunk 4]
```diff
@@ -1 +1,4 @@
 export * from './useDataRef';
```
Comment (LetItRock) [inline packages/react/src/hooks/index.ts]: ✏ **suggestion:** can we please move `useDataRef, useBrowserTabsChannel` to the `src/hooks/internal` folder? Let's not expose these.

File: packages/react/src/components/NovuProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { Novu, NovuOptions } from '@novu/js';
```
Comment (LetItRock) [inline packages/react/src/components/NovuProvider.tsx]: ✏ **suggestion:** Let's keep this file under the `hooks` folder for a better tree-shaking capability.

File: packages/react/src/components/NovuProvider.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import { createContext, useContext, useMemo } from 'react';
+
+type NovuProviderProps = NovuOptions & {
```
Comment (LetItRock) [inline packages/react/src/components/NovuProvider.tsx]: 🗒 **note (non-blocking):** we should probably expose this type

File: packages/react/src/hooks/useWebsocketEvent.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  useEffect(() => {
+    const resolveLock = requestLock(`nv.${webSocketEvent}`, () => {
+      novu.on(webSocketEvent, updateReadCount);
+    });
+
+    return () => {
+      novu.off(webSocketEvent, updateReadCount);
+      resolveLock();
+    };
+  }, []);
```
Comment (LetItRock) [inline packages/react/src/hooks/useWebsocketEvent.ts]: ☑ **chore:** This hook is based on the Web APIs BroadcastChannel and Web Lock, we need to refactor it from these dependencies, like hiding under the `if` statement when these are not available or if not possible remove.

File: packages/react/src/components/NovuProvider.tsx [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  children: JSX.Element;
+};
+
+const NovuContext = createContext<Novu | undefined>(undefined);
+
+export const NovuProvider = (props: NovuProviderProps) => {
```
Comment (LetItRock) [inline packages/react/src/components/NovuProvider.tsx]: ❓ **question:** I don't see that this provider is used internally, do you plan to do this in another pr? if yes, please create the ticket.
The hardest part would be to share it between react/solid.

File: playground/nextjs/src/pages/_app.tsx [hunk 9]
```diff
@@ -1,11 +1,15 @@
...
+import { NovuProvider } from '@novu/react';
 import type { AppProps } from 'next/app';
 import Layout from './layout';
 
 export default function App({ Component, pageProps }: AppProps) {
   return (
-    <Layout>
-      <Component {...pageProps} />
-    </Layout>
+    <NovuProvider {...novuConfig}>
```
Comment (LetItRock) [inline playground/nextjs/src/pages/_app.tsx]: ✏ **suggestion:** Let's wrap only the layout of the page (or route path) where we use hooks, because otherwise, we will see doubled session requests under the network tab when visiting other pages; and I'm not sure if this won't interfere currently;

File: playground/nextjs/src/pages/hooks/index.tsx [hunk 10]
```diff
@@ -0,0 +1,41 @@
+import { useNotifications, usePreferences, useCounts } from '@novu/react';
+
+export default function HooksPage() {
```
Comment (LetItRock) [inline playground/nextjs/src/pages/hooks/index.tsx]: ✏ **suggestion:** It would be nice if we could build the "real" UI with the hooks, something similar to what we have in the Inbox Playground app, just to dogfood this better and see if we cover all main functional cases.

File: packages/react/src/hooks/useNotifications.ts [hunk 11]
```diff
@@ -0,0 +1,130 @@
+import { useState, useEffect, useCallback } from 'react';
```
Comment (LetItRock) [inline packages/react/src/hooks/useNotifications.ts]:
```suggestion
import { useState, useEffect } from 'react';
```

File: packages/react/src/hooks/useNotifications.ts [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+  };
+
+  useEffect(() => {
+    setAfter(undefined);
+    setHasMore(false);
+    fetchNotifications();
+
+    on('notifications.list.updated', sync);
+    on('notifications.list.pending', sync);
+    on('notifications.list.resolved', sync);
```
Comment (LetItRock) [inline packages/react/src/hooks/useNotifications.ts]: ✏ **suggestion:** let's clear all these too

File: packages/react/src/hooks/useNotifications.ts [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    const response = await notifications.list({
+      tags,
+      read,
+      archived,
+      limit,
+      after: after,
+    });
+    if (response.error) {
+      setError(response.error as NovuError);
+      onError?.(response.error as NovuError);
```
Comment (LetItRock) [inline packages/react/src/hooks/useNotifications.ts]:
```suggestion
      setError(response.error);
      onError?.(response.error);
```

File: packages/react/src/hooks/useNotifications.ts [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+    }
+    setData(event.data.notifications);
+    const length = event.data.notifications.length;
+    setAfter(length ? event.data.notifications[length - 1].id : undefined);
+    setHasMore(event.data.hasMore);
+  };
+
+  useEffect(() => {
+    setAfter(undefined);
+    setHasMore(false);
```
Comment (LetItRock) [inline packages/react/src/hooks/useNotifications.ts]: ❓ **question:** call `resetState` ?

File: packages/react/src/hooks/useNotifications.ts [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+  const [hasMore, setHasMore] = useState(false);
+  const [after, setAfter] = useState<string>();
+
+  const sync = (event: { data?: ListNotificationsResponse }) => {
+    if (!event.data) {
+      return;
+    }
+    if (!isSameFilter(event.data.filter, { tags, read, archived })) {
+      return;
+    }
```
Comment (LetItRock) [inline packages/react/src/hooks/useNotifications.ts]:
```suggestion
    if (!event.data || !isSameFilter(event.data.filter, { tags, read, archived })) {
      return;
    }
```

File: packages/react/src/hooks/useNotifications.ts [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+  const fetchMore = async () => {
+    if (!hasMore || isFetching) return;
+    return fetchNotifications();
+  };
+
+  const readAll = async () => {
+    const res = await notifications.readAll({ tags });
+    if (res.error) {
+      setError(res.error);
+    }
```
Comment (LetItRock) [inline packages/react/src/hooks/useNotifications.ts]: ❓ **question:** I'm not sure if we should set the error in the state, because then the question is why we are not calling `onSuccess` and `onError` props, because they are for notifications. So maybe we should just return the result in this case. WDYT?

File: packages/react/src/hooks/usePreferences.ts [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    on('preferences.list.pending', sync);
+    on('preferences.list.resolved', sync);
+  }, []);
+
+  const fetchPreferences = async () => {
+    setIsFetching(true);
+    const response = await preferences.list();
+    if (response.error) {
+      setError(response.error as NovuError);
+      onError?.(response.error as NovuError);
```
Comment (LetItRock) [inline packages/react/src/hooks/usePreferences.ts]:
```suggestion
      setError(response.error);
      onError?.(response.error);
```

File: packages/react/src/hooks/useCounts.ts [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+type UseCountsProps = {
+  filters: NotificationFilter[];
+  onSuccess?: (data: Notification[]) => void;
+  onError?: (error: NovuError) => void;
+};
+
+type CountsData = {
+  count: number;
+  filter: NotificationFilter;
+};
```
Comment (LetItRock) [inline packages/react/src/hooks/useCounts.ts]:
```suggestion
```

File: packages/react/src/hooks/useCounts.ts [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+type UseCountsResult = {
+  counts?: Count[];
+  error?: NovuError;
+  isLoading: boolean; // initial loading
+  isFetching: boolean; // the request is in flight
+  refetch: () => Promise<void>;
+};
+
+export const useCounts = (props: UseCountsProps): UseCountsResult => {
```
Comment (LetItRock) [inline packages/react/src/hooks/useCounts.ts:25]: ❓ **question:** Should we take into account that the `filters` might change?

File: packages/react/src/hooks/useCounts.ts [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    },
+  });
+
+  useEffect(() => {
+    sync();
+  }, []);
+
+  const refetch = async () => {
+    await sync();
+    return;
```
Comment (LetItRock) [inline packages/react/src/hooks/useCounts.ts]:
```suggestion
```

File: packages/react/src/hooks/useCounts.ts [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+import { NotificationFilter, NovuError } from '@novu/js';
+import { useEffect, useState } from 'react';
+import { useNovu } from '../components/NovuProvider';
+import { useWebSocketEvent } from './useWebsocketEvent';
+import { Notification } from '@novu/js';
+
+type UseCountsProps = {
+  filters: NotificationFilter[];
+  onSuccess?: (data: Notification[]) => void;
+  onError?: (error: NovuError) => void;
```
Comment (LetItRock) [inline packages/react/src/hooks/useCounts.ts]: ☑ **chore:** I don't see we call any of these.

File: packages/react/src/hooks/useNotifications.ts [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+  isFetching: boolean;
+  refetch: () => Promise<void>;
+  fetchMore: () => Promise<void>;
+  hasMore: boolean;
+};
+
+export type UseNotificationsProps = {
+  tags?: string[];
+  read?: boolean;
+  archived?: boolean;
```
Comment (LetItRock) [inline packages/react/src/hooks/useNotifications.ts:8]: ❓ **question:** I'm wondering whether by default `archived` should be set to `false` 🤔 wdyt?

**Review by LetItRock — COMMENTED** [review]
1. The after prop is not cleared somehow

https://github.com/user-attachments/assets/b0d8d1c4-e925-4cb4-afe6-3203f5b5403a

2. After archiving notifications the `hasMore` stays as `true`.

https://github.com/user-attachments/assets/5ab31b4d-ca0b-401f-9827-19a97697bb00

3. After refetching notifications the `after` is not cleared.

https://github.com/user-attachments/assets/46611d09-2f53-4402-b977-bcf335dd5345

File: packages/react/src/hooks/useNotifications.ts [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+
+export const useNotifications = (props?: UseNotificationsProps): UseNotificationsResult => {
+  const { tags, read, archived, limit, onSuccess, onError } = props || {};
+  const { notifications, on } = useNovu();
+  const [data, setData] = useState<Array<Notification>>();
+  const [error, setError] = useState<NovuError>();
+  const [isLoading, setIsLoading] = useState(true);
+  const [isFetching, setIsFetching] = useState(false);
+  const [hasMore, setHasMore] = useState(false);
+  const [after, setAfter] = useState<string>();
```
Comment (LetItRock) [inline packages/react/src/hooks/useNotifications.ts]: ❓ **question:** Do we need to have it in the state? we can calculate it on the fly.

File: packages/react/src/components/NovuProvider.tsx (hunk 8 above)
Comment (desiprisg) [reply packages/react/src/components/NovuProvider.tsx]: This is meant to be in react. If we decide to share this with the `<Inbox />`, we would pass the novu instance in the `mountComponent`  method.

File: packages/react/src/hooks/index.ts (hunk 4 above)
Comment (desiprisg) [reply packages/react/src/hooks/index.ts]: They are not exposed right now. I explicitly export the hooks that we want to expose.

File: playground/nextjs/src/pages/hooks/index.tsx (hunk 10 above)
Comment (desiprisg) [reply playground/nextjs/src/pages/hooks/index.tsx]: Nice idea but I'd say we should do this in another PR, and maybe include in a starter repo as well.

File: packages/react/src/hooks/useNotifications.ts (hunk 15 above)
Comment (desiprisg) [reply packages/react/src/hooks/useNotifications.ts]: I kinda find it easier to read this way, but I'll change it :P

File: packages/react/src/hooks/useNotifications.ts (hunk 16 above)
Comment (desiprisg) [reply packages/react/src/hooks/useNotifications.ts]: Hmm, yeah you're right

File: packages/react/src/hooks/useNotifications.ts (hunk 22 above)
Comment (desiprisg) [reply packages/react/src/hooks/useNotifications.ts:8]: It seems kind of misleading to me if it has a default value.

File: packages/react/src/hooks/useNotifications.ts (hunk 23 above)
Comment (desiprisg) [reply packages/react/src/hooks/useNotifications.ts]: Nice suggestion!

File: packages/react/src/hooks/useWebsocketEvent.ts (hunk 7 above)
Comment (desiprisg) [reply packages/react/src/hooks/useWebsocketEvent.ts]: Is this just for server env or do you think it applies elsewere as well ?
Also, what do we want the behaviour to be in that case?

Comment (LetItRock) [reply packages/react/src/hooks/useWebsocketEvent.ts]: Ideally, for the `web` env, we want to keep it, for the `mobile` or `server` I think we should hide (if possible under if) it as it won't be available.

File: packages/react/src/components/NovuProvider.tsx (hunk 8 above)
Comment (LetItRock) [reply packages/react/src/components/NovuProvider.tsx]: Let's have a ticket to improve this, like: "Use the NovuProvide internally in the Inbox react" or smth

File: playground/nextjs/src/pages/hooks/index.tsx (hunk 10 above)
Comment (LetItRock) [reply playground/nextjs/src/pages/hooks/index.tsx]: ok, I'll create a ticket

File: packages/react/src/hooks/useNotifications.ts (hunk 22 above)
Comment (LetItRock) [reply packages/react/src/hooks/useNotifications.ts:8]: I think that in most cases devs who will build using hooks will use it to fetch `read/unread` notifications, that's why I'm saying to make `archived: false` by default.

File: packages/react/src/hooks/useWebsocketEvent.ts (hunk 7 above)
Comment (desiprisg) [reply packages/react/src/hooks/useWebsocketEvent.ts]: I added it by checking if the lock api exists in this case.

File: packages/react/src/components/NovuProvider.tsx (hunk 8 above)
Comment (desiprisg) [reply packages/react/src/components/NovuProvider.tsx]: I think this is resolved since we will go with explicit or implicit novu.

File: packages/react/src/hooks/NovuProvider.tsx [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+
+type NovuProviderProps = NovuOptions & {
+  children: JSX.Element;
+};
+
+const NovuContext = createContext<Novu | undefined>(undefined);
+
+export const NovuProvider = (props: NovuProviderProps) => {
+  const { children, ...options } = props;
+  const novu = useMemo(() => new Novu(options), []);
```
Comment (LetItRock) [inline packages/react/src/hooks/NovuProvider.tsx]: We should probably create a new Novu instance when the props change.

File: packages/react/src/hooks/index.ts [hunk 25]
```diff
@@ -1 +1,3 @@
-export * from './useDataRef';
+export * from './useNotifications';
+export * from './usePreferences';
+export * from './useCounts';
```
Comment (LetItRock) [inline packages/react/src/hooks/index.ts:3]: ✏ **suggestion:** we should export here the provider as well and only `useNovu` hook (not exported rn)

File: packages/react/src/hooks/index.ts [hunk 26]
```diff
@@ -1 +1,3 @@
-export * from './useDataRef';
+export * from './useNotifications';
```
Comment (LetItRock) [inline packages/react/src/hooks/index.ts:1]: ☑ **chore:** btw I'm missing the `package.json` export changes for the hooks, please check how we do this in the js package and do it similarly

File: packages/react/src/hooks/NovuProvider.tsx [hunk 27]
```diff
@@ -0,0 +1,30 @@
+import { Novu, NovuOptions } from '@novu/js';
+import { createContext, useContext, useMemo } from 'react';
+
+type NovuProviderProps = NovuOptions & {
+  children: JSX.Element;
+};
+
+const NovuContext = createContext<Novu | undefined>(undefined);
+
+export const NovuProvider = (props: NovuProviderProps) => {
```
Comment (LetItRock) [inline packages/react/src/hooks/NovuProvider.tsx]: ❓ **question:** I'm not sure what we should do with the props on the Inbox in this case:
```
<NovuProvider subscriberId="subscriberId" applicationIdentifier="applicationIdentifier">
   <Inbox subscriberId="subscriberId" applicationIdentifier="applicationIdentifier" />
</NovuProvider>
```

File: packages/react/src/hooks/internal/useBrowserTabsChannel.ts [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+  onMessage,
+}: {
+  channelName: string;
+  onMessage: (args: T) => void;
+}) => {
+  const [tabsChannel] = useState(new BroadcastChannel(channelName));
+
+  const postMessage = (data: T) => {
+    tabsChannel.postMessage(data);
+  };
```
Comment (LetItRock) [inline packages/react/src/hooks/internal/useBrowserTabsChannel.ts]: ☑ **chore:** this function might still fail in RN, it's exposed and used when the ws event comes

File: packages/react/src/components/Inbox.tsx [hunk 29]
```diff
@@ -35,8 +36,30 @@ const DefaultInbox = (props: DefaultInboxProps) => {
...
+    <NovuProvider
+      applicationIdentifier={applicationIdentifier}
+      subscriberId={subscriberId}
+      subscriberHash={subscriberHash}
+      backendUrl={backendUrl}
+      socketUrl={socketUrl}
+    >
+      <InboxChild {...props} />
+    </NovuProvider>
+  );
```
Comment (LetItRock) [inline packages/react/src/components/Inbox.tsx:56]: Support hooks in the Inbox children components.

File: packages/react/src/hooks/useNotifications.ts (hunk 22 above)
Comment (LetItRock) [reply packages/react/src/hooks/useNotifications.ts:8]: we agreed on setting it to `archive: false` by default: https://novu.slack.com/archives/C06TPLSQGUT/p1725610768168889

**Review by LetItRock — APPROVED** [review]

**Review by BiswaViraj — CHANGES_REQUESTED** [review]
```
Error: This JSX tag's 'children' prop expects a single child of type 'Element', but multiple children were provided.ts(2746)

<NovuProvider {...novuConfig}>
  <div>hello</div>
  <Notifications />
</NovuProvider>
      
```
It is not accepting multiple children

File: packages/react/src/hooks/NovuProvider.tsx [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+export const useNovu = () => {
+  const context = useContext(NovuContext);
+  if (!context) {
+    throw new Error('useNovu must be used within a <NovuProvider />');
+  }
+
+  return context;
+};
+
+export const useUnsafeNovu = () => {
```
Comment (SokratisVidros) [inline packages/react/src/hooks/NovuProvider.tsx:36]: 💯

File: packages/react/src/hooks/index.ts (hunk 26 above)
Comment (SokratisVidros) [reply packages/react/src/hooks/index.ts:1]: @LetItRock is this comment still pending or can we resolve it?

File: packages/react/src/utils/requestLock.ts [hunk 31]
```diff
@@ -0,0 +1,28 @@
+export function requestLock(id: string, cb: (id: string) => void) {
```
Comment (SokratisVidros) [inline packages/react/src/utils/requestLock.ts:1]: ❓ Did we add locks support now or did we have it already in @novu/js? Actually I am a bit surprised this is not in @novu/js

**Review by SokratisVidros — APPROVED** [review]
🪝

File: packages/react/src/hooks/index.ts (hunk 26 above)
Comment (LetItRock) [reply packages/react/src/hooks/index.ts:1]: @SokratisVidros we can resolve it, the exports were added

File: packages/react/src/utils/requestLock.ts (hunk 31 above)
Comment (LetItRock) [reply packages/react/src/utils/requestLock.ts:1]: @SokratisVidros We have it for the JS package. However, we needed to copy this code for the `useWebSocketEvent` hook implementation. But we can actually expose it from the JS package as part of the `@novu/js/internal` export or smth similar.

**Review by BiswaViraj — APPROVED** [review]
